### PR TITLE
Confirm unsaved code component changes

### PR DESCRIPTION
**Pull request by apedroferreira**
Partial PR from: https://github.com/mui/mui-toolpad/pull/1618

- [x]  Confirm when leaving code component editor with unsaved changes

For the code component editor, because there are several ways to leave the page, I created a new `hasUnsavedChanges` state in the same app state where view changes are controlled (where other state already was such as `undoStack`, `redoStack`, `selectedNodeId`, `currentView` and `currentTab`).

Also, because so far all these "app state" properties have been in the same state/file as the DOM loading properties and the DOM save state, but they sound like different concepts, I changed the `DomLoader` file where this state was in to `AppState` and tried to split things a bit inside that file. This included splitting the `domApi` into separate `domApi` and `appStateApi`.
Of course this is just a suggestion and it's up for discussion, or we can use it as the starting point for a bigger separation.

Preview: https://toolpad-pr-1628.onrender.com/

### CODE REVIEW

File: packages/toolpad-app/src/toolpad/AppState.tsx [hunk 1]
```diff
@@ -17,18 +17,21 @@ import { hasFieldFocus } from '../utils/fields';
 import { DomView, getViewFromPathname } from '../utils/domView';
 import config from '../config';
 
+export function getNodeHashes(dom: appDom.AppDom): NodeHashes {
+  return mapValues(dom.nodes, (node) => insecureHash(JSON.stringify(node)));
+}
+
 export type ComponentPanelTab = 'component' | 'theme';
 
-export type DomAction =
```
Comment (apedroferreira) [inline packages/toolpad-app/src/toolpad/AppState.tsx:22]: It might be better to just see how this file would look now instead of just looking at the diff.

**Comment by prakhargupta1** [comment]
@apedroferreira Could you please trigger a re-deployment of this branch?

**Comment by apedroferreira** [comment]
> @apedroferreira Could you please trigger a re-deployment of this branch?

Redeploying, I think.

This is ready to review btw, most of the changes are just suggested method name changes but I can easily revert those, was just trying to improve the API because it was getting a bit messy and I didn't want to make it even worse :/

**Comment by Janpot** [comment]
These sort of big changes are going to be unpleasant to merge back into the direction 13 branch.

**Review by Janpot — APPROVED** [review]

**Comment by apedroferreira** [comment]
> These sort of big changes are going to be unpleasant to merge back into the direction 13 branch.

Yeah, I can merge it to direction 13 too then already and take care of the conflicts if there are any. Thanks for checking this!

**Comment by Janpot** [comment]
Don't worry, I went over it again, I may have exaggerated a bit, it should be fine.

**Pull request merged**